OPENNLP-880: Refactor data indexer

danielruss opened the pull request:
Added a patch that changes multiple items (Sorry That I did not fix 1 item at a time-I am a scientist not a software engineer):

1.) Added method to EventTrainer – train(DataIndexer indexer) –Then I added the method to AbstractEventTrainer. No other code needed to be changed.
2.) Created a new class – PluggableParmeters: Only AbstractTrainers had access to getXXXParam(Value,Default). So I pulled out this functionality into a separate class. Now both AbstractTrainers and AbstractDataIndexers can hold parameters.
3.) Refactored DataIndexer. This touched a lot of code. Added init(Map,Map) method and index(ObjectStream<Event>) method. Changed 1-Pass and 2-Pass DataIndexers. Everywhere 1-pass/2-pass indexers where created, I changed the constructor, and added the init and index methods.
4.) Changed GIS.doTrain(indexer) to use the parameters passed in the init method.
5.) QNTrainer – created a working init method. Changed the isValid method so it is not the init method.

#4 Make the GIS class work within the new training API

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java:
```diff
@@ -39,141 +39,175 @@
+		boolean sort=parameters.getBooleanParam(SORT_PARAM, SORT_DEFAULT);
+		
+		Map<String, Integer> predicateIndex = new HashMap<>();
+		LinkedList<Event> events;
```
use the interface List on LHS

danielruss replied:
This is code that was originally in the constructor that I moved into the index method.  I will be happy to change it to a List<Event>.

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java:
```diff
@@ -39,141 +39,175 @@
+	 *          an <code>int</code> value
+	 * @return a <code>TLinkedList</code> value
+	 */
+	private LinkedList<Event> computeEventCounts(ObjectStream<Event> eventStream,
```
change method signature to return interface List as opposed to a concrete class

danielruss replied:
Not a problem, I did not change this method, but I will have it return a List<Event>.

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java:
```diff
@@ -39,141 +39,175 @@
+			Map<String, Integer> predicatesInOut, int cutoff) throws IOException {
+		Set<String> predicateSet = new HashSet<>();
+		Map<String, Integer> counter = new HashMap<>();
+		LinkedList<Event> events = new LinkedList<>();
```
change to List<> on LHS

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java:
```diff
@@ -39,141 +39,175 @@
+	 *          order to be included in the model.
+	 */
+	@Deprecated
+	public OnePassDataIndexer(ObjectStream<Event> eventStream, int cutoff, boolean sort)
```
if Deprecated, do we wanna just not add it ?

danielruss replied:
I didn't add this, After deprecating the methods, I moved them to the bottom of the file.

kottmann replied:
- Moving code makes it harder to review, it could have stayed where it was, then it would just be a matter of one line insert for the deprecation.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/ml/PluggableParameters.java:
```diff
@@ -0,0 +1,67 @@
+		if (reportMap != null)
+			reportMap.put(key, valueString);
+
+		return valueString;
```
All our code uses two spaces for an indent and not tabs. Could you please reformat that accordingly?

We have that explained here:
https://opennlp.apache.org/code-conventions.html

kottmann commented:
- I get a compilation error in this branch:
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/joern/dev/opennlp/opennlp-tools/src/test/java/opennlp/tools/ml/MockEventTrainer.java:[27,8] opennlp.tools.ml.MockEventTrainer is not abstract and does not override abstract method train(opennlp.tools.ml.model.DataIndexer) in opennlp.tools.ml.EventTrainer
[INFO] 1 error

- There are also still many changes with tab indents instead of two spaces.

danielruss commented:
I made the changes to the method signatures as requested by smathi, but we missed the change to the signature to OnePassRealValueEventStream.  Even though the method signatures are similar, they are different so index was not overridden.  I changed the signature and added the "@Override". If the annotation was on the method before, Eclipse would have caught that the signature needed to be changed.

danielruss commented:
sorry about that... hit the wrong button.

smarthi commented:
@danielruss could u update this PR with the present codebase, its got merge conflicts that need to be resolved. thanks!

danielruss commented:
had a major merging issue.  Took a while to fix.  I had to open a new branch and write over the old one (it appeared that I wrote brat-annotation file)  The history appears correct now, though I lost the comments smarthi and kottmann made.

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/AbstractEventTrainer.java:
```diff
@@ -61,26 +64,38 @@ public boolean isValid() {
-    String dataIndexerName = getStringParam(DATA_INDEXER_PARAM,
+    String dataIndexerName = parameters.getStringParam(DATA_INDEXER_PARAM,
         DATA_INDEXER_TWO_PASS_VALUE);
+    Map<String,String> indexParams=new HashMap<String, String>();
```
nitpick: leave a space on both sides of '='

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java:
```diff
@@ -39,141 +39,174 @@
+	public OnePassDataIndexer(){}
+
+	@Override
+	public void index(ObjectStream<Event> eventStream) throws IOException{
```
Indentation is incorrect in this class, should be two spaces and not tab.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/AbstractDataIndexer.java:
```diff
@@ -33,6 +35,18 @@
+  public static final String SORT_PARAM = "sort";
+  public static final boolean SORT_DEFAULT = true;
+
+  PluggableParameters parameters;
```
Can this be protected? Should only be used by sub-classes.

kottmann commented:
What do you think about using TrainingParameters instead of adding a new class doing kind of the same? Maybe it is not a good idea to do that right now, but we should maybe open an issue for that.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/ml/AbstractEventTrainer.java:
```diff
@@ -92,8 +107,8 @@ public final MaxentModel train(ObjectStream<Event> events) throws IOException {
 
     MaxentModel model = doTrain(indexer);
```
Should we not just call the new train method from here?

danielruss replied:
Before we call the doTrain(), we still need to check to make sure the training parameters are valid.  The isValid() method in AbstractEventTrainer only check if the dataindexer is 1 or two pass, which is currently useful.  However, every trainer should override the isValid() method with a more specific method.  Either way, if you train with indexedData, you should still check if the parameters are valid.

kottmann replied:
I mean the train(ObjectStream<Event> events) could call train(DataIndexer indexer) to do the actual training, otherwise the code is kind of duplicated.

danielruss replied:
yes, you are correct.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/ml/PluggableParameters.java:
```diff
@@ -0,0 +1,67 @@
+    }
+  }
+
+  public Map<String, String> getReportMap() {
```
Not sure we should add this here. This class should only serve one purpose.

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/maxent/GISTrainer.java:
```diff
@@ -220,7 +223,12 @@ public void setGaussianSigma(double sigmaValue) {
   public GISModel trainModel(ObjectStream<Event> eventStream, int iterations, int cutoff) throws IOException {
-    return trainModel(iterations, new OnePassDataIndexer(eventStream,cutoff),cutoff);
+	DataIndexer indexer=new OnePassDataIndexer();
+	Map<String, String> params=new HashMap<String, String>();
```
no need to specify types on RHS

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/maxent/GISTrainer.java:
```diff
@@ -220,7 +223,12 @@ public void setGaussianSigma(double sigmaValue) {
+	Map<String, String> params=new HashMap<String, String>();
+	params.put(GIS.ITERATIONS_PARAM, Integer.toString(iterations));
+	params.put(GIS.CUTOFF_PARAM, Integer.toString(cutoff));
+	indexer.init(params, new HashMap<String, String>());
```
no need to specify types in new HashMap<>()

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/AbstractDataIndexer.java:
```diff
@@ -33,6 +35,18 @@
+
+  PluggableParameters parameters;
+
+  public void init(Map<String,String> indexingParameters,Map<String, String> reportMap){
```
leave a space between the comma and next parameter, change to 

`indexingParameters, Map<String, String> `

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassRealValueDataIndexer.java:
```diff
@@ -72,16 +60,17 @@ protected int sortAndMerge(List<ComparableEvent> eventsToCompare,boolean sort) t
 
+  @Override
+  protected List<ComparableEvent> index(List<Event> events, Map<String,Integer> predicateIndex) {
+    Map<String,Integer> omap = new HashMap<String,Integer>();
```
no need to specify types on RHS - i would fix this throughout this PR

danielruss commented:
I agree with Jörn's comment about PluggableParameters.  I will move the get(Int/String/Boolean)Parameters into TrainingParameters

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/maxent/GIS.java:
```diff
@@ -162,7 +162,7 @@ public static GISModel trainModel(ObjectStream<Event> eventStream, int iteration
   public static GISModel trainModel(ObjectStream<Event> eventStream, int iterations,
       int cutoff, boolean smoothing, boolean printMessagesWhileTraining)
-      throws IOException {
+          throws IOException {
```
Please avoid tab indents, replace with 2 spaces instead

smarthi commented on opennlp-tools/src/main/java/opennlp/tools/ml/model/RealValueFileEventStream.java:
```diff
@@ -110,15 +110,15 @@ public static void main(String[] args) throws IOException {
       params.put(AbstractTrainer.CUTOFF_PARAM, args[ai++]);
     }
-    PluggableParameters parameters=new PluggableParameters(params, new HashMap<String, String>());
+    PluggableParameters parameters=new PluggableParameters(params, new HashMap<>());
```
add spaces around '='

smarthi commented:
@danielruss sorry, but could u please update this PR again against present codebase? Now that we have checkstyle enabled you may want to fix any code style issues that are flagged.

kottmann commented:
@danielruss this is good to merge code changes are all good and I did also run a couple of tests to ensure it doesn't break anything. After you resolved smarthi comment from above we can get it finally merged.

We will need to squash all your commits, you can do that by yourself if you want and you probably need to rebase to see the checkstyle work we did.

Please also notice we switched from trunk to master branch.

kottmann commented:
@danielruss do you want to squash to yourself, or should I? We are ready to merge your work.

danielruss commented:
@kottmann I am not sure how to do this.  Do you mind squashing the comments? I'll try to find it in the manual.

kottmann commented:
Thanks a lot for helping with this! And thanks for your patience with our bigger code changes in between.
